packages/alpinejs/src/directives/x-for.js

```diff
@@ -30,49 +32,131 @@
+        const lookup = templateEl._x_lookup;
+        const prevKeys = templateEl._x_prevKeys;
 
-        let lookup = el._x_lookup
-        let prevKeys = el._x_prevKeys
-        let scopes = []
-        let keys = []
+        let lastEl = templateEl;
+        keys.forEach((key, index) => {
+            const hydrationTarget = hydratables[key];
+            if (!hydrationTarget) return;
+            const scope = scopes[index];
+            lookup[key] = hydrationTarget;
+            prevKeys.push(key);
 
-        // In order to preserve DOM elements (move instead of replace)
-        // we need to generate all the keys for every iteration up
-        // front. These will be our source of truth for diffing.
-        if (isObject(items)) {
-            items = Object.entries(items).map(([key, value]) => {
-                let scope = getIterationScopeVariables(iteratorNames, value, key, items)
+            addScopeToNode(hydrationTarget, reactive(scope), templateEl);
+            hydrationTarget._x_forScope = hydrationTarget._x_dataStack[0];
 
-                evaluateKey(value => keys.push(value), { scope: { index: key, ...scope} })
+            if ('morph' in Alpine)
+            mutateDom(() => {
+                hydrationTarget.removeAttribute('key');
+                Alpine.morph(
+                hydrationTarget,
+                document.importNode(templateEl.content, true).firstElementChild
+                    .outerHTML
+                );
+                initTree(hydrationTarget);
+            });
+
+            lastEl.after(hydrationTarget);
+            lastEl = hydrationTarget;
+        });
+    });
+};
+
+function getScopesFromItems (
+    items,
+    iteratorNames,
+    evaluateKey
+) {
+    const scopes = [];
+    const keys = [];
+
+    // Support number literals. Ex: x-for="i in 100"
+    if (isNumeric(items) && items >= 0)
+      items = Array.from({ length: items }, (_, i) => i + 1);
+
+    if (items === undefined) items = [];
+
+    // In order to preserve DOM elements (move instead of replace)
+    // we need to generate all the keys for every iteration up
+    // front. These will be our source of truth for diffing.
+    if (isObject(items)) {
+      items = Object.entries(items).map(([key, value]) => {
+        const scope = getIterationScopeVariables(
+          iteratorNames,
+          value,
+          key,
+          items
+        );
+
+        evaluateKey((value) => keys.push(value), {
+          scope: { index: key, ...scope },
+        });
+
+        scopes.push(scope);
+      });
+    } else {
+        items.forEach((item, index) => {
+            const scope = getIterationScopeVariables(
+            iteratorNames,
+            item,
+            index,
+            items
+            );
+
+            evaluateKey((value) => keys.push(value), {
+            scope: { index, ...scope },
+            });
+
+            scopes.push(scope);
+        });
+    }
+    return [scopes, keys];
+};
```

Comment:
**Suggestion:** The `getScopesFromItems` function does not handle the case when the `items` parameter is not an object or an array. This could lead to unexpected behavior or errors. It would be better to add a condition to check if the `items` parameter is an object or an array and handle the case accordingly.
```suggestion
function getScopesFromItems (
    items,
    iteratorNames,
    evaluateKey
) {
    const scopes = [];
    const keys = [];

    // Support number literals. Ex: x-for="i in 100"
    if (isNumeric(items) && items >= 0)
      items = Array.from({ length: items }, (_, i) => i + 1);

    if (items === undefined) items = [];

    // In order to preserve DOM elements (move instead of replace)
    // we need to generate all the keys for every iteration up
    // front. These will be our source of truth for diffing.
    if (isObject(items)) {
      items = Object.entries(items).map(([key, value]) => {
        const scope = getIterationScopeVariables(
          iteratorNames,
          value,
          key,
          items
        );

        evaluateKey((value) => keys.push(value), {
          scope: { index: key, ...scope },
        });

        scopes.push(scope);
      });
    } else if (Array.isArray(items)) {
        items.forEach((item, index) => {
            const scope = getIterationScopeVariables(
            iteratorNames,
            item,
            index,
            items
            );

            evaluateKey((value) => keys.push(value), {
            scope: { index, ...scope },
            });

            scopes.push(scope);
        });
    } else {
        warn('The items parameter should be an object or an array.');
    }
    return [scopes, keys];
};
```